### ISSUE TITLE
[CSA] False negative: null dereference report recovered by loop widening

### ISSUE THREAD

**Issue by lynae99** [llvm/llvm-project#214230]
Hi, I found a false negative in `core.NullDereference` loop path exploration.

## Program

```c
#include <stddef.h>

int check(int count) {
  int *local_ptr = NULL;
  for (int i = 0; i < count; ++i) {
    if (local_ptr == NULL && i > 5)
      return *local_ptr;
  }
  return 0;
}
```

## Actual result

CSA emits no `core.NullDereference` warning by default:

```sh
clang --analyze \
  -Xclang -analyzer-checker=core.NullDereference \
  loop_widening_null_dereference.c
```

Enabling loop widening reports the dereference:

```sh
clang --analyze \
  -Xclang -analyzer-checker=core.NullDereference \
  -Xclang -analyzer-config -Xclang widen-loops=true \
  loop_widening_null_dereference.c
```

The path `count >= 7` is feasible, `local_ptr` is still null, and `i > 5` makes the dereference definite.

## Runtime confirmation

Building the provided file with `-DRUNTIME -fsanitize=address` and executing `check(7)` produces an AddressSanitizer null-read crash.

This does not appear to be the inlined defensive-check suppression heuristic: `-analyzer-config suppress-null-return-paths=false` does not recover the diagnostic. The report is recovered specifically by `widen-loops=true`.

## Expected result

It would be useful for the default loop exploration policy to preserve enough state to diagnose this feasible null-dereference path, or to make the suppression tradeoff explicit.

## Verification

```text
clang version 24.0.0git (https://github.com/llvm/llvm-project.git 1cb7e838cd47ecad4050948c0c907ecb1f466ac3)
```

**Comment by NagyDonat** [comment]
Yes, it is known that loop widening can find some results that cannot be found otherwise. However, it also has drawbacks: it probably increases the analysis time and it performs invalidation, which can cause false positives [1].

We have vague plans to evaluate the advantages and drawbacks of the loop widening mode, and we will enable it if the benefits are more relevant than the costs. However I don't know when will this happen – but until then anybody can enable the loop widening mode if it is useful for them.

Unfortunately I don't think that new contributors can effectively help in this area: evaluating the performance and the result quality will probably require some expertise and a complex test environment. (However, if somebody is willing to invest significant effort into this, then feel free to contact me.)

----------

[1] "invalidation" is the process when the analyzer discards information about values that may have been changed in the loop. This is needed because blindly using the old values would be a clear error (producing lots of false positives) but discarding information also produces false positives with the following pattern:
- The analyzer makes an assumption about a value `x` (e.g. that it is non-null).
- This assumption is used in some way (e.g. we allocate something if `x` is non-null).
- The loop widening happens and it invalidates the knowledge about `x` (because `x` _may have been written_ by the loop).
- The analyzer makes a different assumption about the value of `x` (e.g. that it is null).
- This assumption is used (e.g. the `free()` call is skipped when `x` is null) which leads to a false positive report.

In C/C++ it is very difficult to accurately find all the values that may be changed in a loop, because there are many ways to indirectly change values.
